src/main/java/com/amazonaws/encryptionsdk/internal/DecryptionHandler.java

```diff
@@ -255,7 +303,27 @@ public ProcessingSummary processBytes(final byte[] in, final int off, final int
 
         int totalParsedBytes = 0;
         if (ciphertextHeadersParsed_ == false) {
-            totalParsedBytes += ciphertextHeaders_.deserialize(bytesToParse, 0);
+            // if maxHeaderSize_ is set, we MUST NOT parse a header with length greater than that value.
+            byte[] headerBytesToParse = maxHeaderSize_ == null || headerBytesParsed_ + bytesToParse.length < maxHeaderSize_
+                    ? bytesToParse
+                    : Arrays.copyOfRange(bytesToParse, 0, maxHeaderSize_ - (int)headerBytesParsed_); // TODO it should be safe to convert this to int, but we might want a stronger assurance
```

Comment:
It might be safer to do the conversion a line above this and have a fallback/ throw if we can't convert to `int`